test/test_ops.py

```diff
@@ -487,6 +487,685 @@ def script_func(input, rois):
+                if roi[0] != n:
+                    continue
+                roi[1:] = (roi[1:] * spatial_scale) - 0.5
+                c_in = 0
+                roi_height = float(roi[4].item() - roi[2].item())
+                roi_width = float(roi[3].item() - roi[1].item())
+                bin_h, bin_w = roi_height / float(pool_h), roi_width / float(pool_w)
+                for c_out in range(0, num_output_channels):
+                    for j in range(0, pool_h):
+                        start_h = float(j) * bin_h + roi[2].item()
+
+                        for i in range(0, pool_w):
+                            start_w = float(i) * bin_w + roi[1].item()
+
+                            roi_bin_grid_h = sampling_ratio if sampling_ratio > 0 else int(np.ceil(roi_height / pool_h))
+                            roi_bin_grid_w = sampling_ratio if sampling_ratio > 0 else int(np.ceil(roi_width / pool_w))
+
+                            val = 0.
+                            for iy in range(0, roi_bin_grid_h):
+                                y = start_h + (iy + 0.5) * bin_h / float(roi_bin_grid_h)
+                                for ix in range(0, roi_bin_grid_w):
+                                    x = start_w + (ix + 0.5) * bin_w / float(roi_bin_grid_w)
+                                    val += bilinear_interpolate(
+                                        in_data[n, c_in, :, :].flatten(),
+                                        in_data.size(-2),
+                                        in_data.size(-1),
+                                        y, x
+                                    )
+                            count = roi_bin_grid_h * roi_bin_grid_w
+                            out_data[r, c_out, j, i] = val / count
+                            c_in += 1
+        return out_data
+
+    def test_ps_roi_align_basic_cpu(self):
+        device = torch.device('cpu')
+        pool_size = 3
+        x = torch.rand(1, 2 * (pool_size ** 2), 7, 7, dtype=self.dtype, device=device)
+        rois = torch.tensor([[0, 0, 0, 5, 5]],  # format is (xyxy)
+                            dtype=self.dtype, device=device)
+
+        pool_h, pool_w = (pool_size, pool_size)
+        ps_roi_align = ops.PSRoIAlign((pool_h, pool_w), spatial_scale=1, sampling_ratio=2)
+        y = ps_roi_align(x, rois)
+
+        gt_y = self.slow_ps_roi_align(x, rois, pool_h, pool_w, device,
+                                      spatial_scale=1, sampling_ratio=2,
+                                      dtype=self.dtype)
+        assert torch.allclose(gt_y, y), 'PSRoIAlign layer incorrect on CPU'
+
+        y = ps_roi_align(x.permute(0, 1, 3, 2), rois)
+        gt_y = self.slow_ps_roi_align(x.permute(0, 1, 3, 2), rois, pool_h, pool_w, device,
+                                      spatial_scale=1, sampling_ratio=-1,
+                                      dtype=self.dtype)
+        assert torch.allclose(gt_y, y), 'PSRoIAlign layer incorrect on CPU'
+
+    def test_ps_roi_align_cpu(self):
```

Comment:
For a follow-up PR: we should simplify the testing for the `roi_*` layers, they can be refactored so that we have many less lines of code.